tempesta_fw/http_msg.c

```diff
@@ -868,17 +869,18 @@ tfw_http_msg_alloc(int type)
 	if (!hm)
 		return NULL;
 
-	hm->h_tbl = (TfwHttpHdrTbl *)tfw_pool_alloc(hm->pool, TFW_HHTBL_SZ(1));
-	if (unlikely(!hm->h_tbl)) {
-		TFW_WARN("Insufficient memory to create message\n");
-		tfw_pool_destroy(hm->pool);
-		return NULL;
+	if (intent != HTTP_MSG_LIGHT) {
+		hm->h_tbl = (TfwHttpHdrTbl *)tfw_pool_alloc(hm->pool, TFW_HHTBL_SZ(1));
+		if (unlikely(!hm->h_tbl)) {
+			TFW_WARN("Insufficient memory to create message\n");
+			tfw_pool_destroy(hm->pool);
+			return NULL;
+		}
+		hm->h_tbl->size = __HHTBL_SZ(1);
+		hm->h_tbl->off = TFW_HTTP_HDR_RAW;
+		memset(hm->h_tbl->tbl, 0, __HHTBL_SZ(1) * sizeof(TfwStr));
```

Comment:
These changes leave a strange feeling of being half-done. On one hand, the new argument allows for creating of special messages without the header table that is required for the parser. On the other hand, the parser is still initialized regardless of that in the code that follows. Then, there's `tfw_http_msg_alloc_err_resp()` that basically does the same thing but it's still there and not removed.

Now the new argument is required for all allocations, and it looks somewhat ugly.

I believe this (and related) code should be refactored. The original top-level interface should stay the same in the form of `tfw_http_msg_alloc(int type)`. A different interface should be introduced for messages that are not going through the parser AND are not modified - which are the messages that are fully prepared from scratch by Tempesta. Perhaps, internally, these two different sets of interfaces may use the same common code - either by calling some inline functions or with macros. `tfw_http_msg_alloc_err_resp()` would be superseded with the new interface.

Perhaps, you could call it `tfw_http_msg_alloc_light()` - nothing better comes to mind right now.

Reply:
Done.